verifiers/internal/gcb/provenance.go

```diff
@@ -0,0 +1,374 @@
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa01 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.1"
+	dsselib "github.com/secure-systems-lab/go-securesystemslib/dsse"
+
+	serrors "github.com/slsa-framework/slsa-verifier/errors"
+	"github.com/slsa-framework/slsa-verifier/options"
+	"github.com/slsa-framework/slsa-verifier/verifiers/internal/gcb/keys"
+)
+
+var GCBBuilderIDs = []string{"https://cloudbuild.googleapis.com/GoogleHostedWorker@v0.2"}
+
+type v01IntotoStatement struct {
+	intoto.StatementHeader
+	Predicate slsa01.ProvenancePredicate `json:"predicate"`
+}
+
+type provenance struct {
+	Build struct {
+		// TODO: compare to verified provenance.
+		// IntotoStatement v01IntotoStatement `json:"intotoStatement"`
+	} `json:"build"`
+	Kind        string           `json:"kind"`
+	ResourceURI string           `json:"resourceUri"`
+	Envelope    dsselib.Envelope `json:"envelope"`
+}
+
+type gloudProvenance struct {
+	ImageSummary struct {
+		Digest               string `json:"digest"`
+		FullyQualifiedDigest string `json:"fully_qualified_digest"`
+		Registry             string `json:"registry"`
+		Repsitory            string `json:"repository"`
+	} `json:"image_summary"`
+	ProvenanceSummary struct {
+		Provenance []provenance `json:"provenance"`
+	} `json:"provenance_summary"`
+}
+
+type GCBProvenance struct {
+	gcloudProv                    *gloudProvenance
+	verifiedProvenance            *provenance
+	verifiedIntotoStatementStruct *v01IntotoStatement
+}
+
+func ProvenanceFromBytes(payload []byte) (*GCBProvenance, error) {
+	var prov gloudProvenance
+	err := json.Unmarshal(payload, &prov)
+	if err != nil {
+		return nil, fmt.Errorf("json.Unmarshal: %w", err)
+	}
+
+	return &GCBProvenance{
+		gcloudProv: &prov,
+	}, nil
+}
+
+func signatureAsRaw(s string) ([]byte, error) {
+	s = strings.ReplaceAll(s, "-", "+")
+	s = strings.ReplaceAll(s, "_", "/")
+	sig, err := base64.StdEncoding.DecodeString(s)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, err.Error())
+	}
+	return sig, nil
+}
+
+func payloadFromEnvelope(env *dsselib.Envelope) ([]byte, error) {
+	payload, err := base64.StdEncoding.DecodeString(env.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, err.Error())
+	}
+	return payload, nil
+}
+
+func (self *GCBProvenance) isVerified() error {
+	// Check that the signature is verified.
+	if self.verifiedIntotoStatementStruct == nil ||
+		self.verifiedProvenance == nil {
+		return serrors.ErrorNoValidSignature
+	}
+	return nil
+}
+
+func (self *GCBProvenance) GetVerifiedIntotoStatement() ([]byte, error) {
+	if err := self.isVerified(); err != nil {
+		return nil, err
+	}
+	d, err := json.Marshal(self.verifiedIntotoStatementStruct)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, err.Error())
+	}
+	return d, nil
+}
+
+func (self *GCBProvenance) VerifyMetadata(provenanceOpts *options.ProvenanceOpts) error {
```

Comment:
some doc comments on what this is doing would be helpful. Same with other Verify* methods.

Reply:
Thank you, done.